Source/NSGridView.m

```diff
@@ -0,0 +1,1280 @@
+      if (ri > 0 && [colWidths count] == 0)
+        {
+          if (rect.size.width > 0.0)
+            {
+              [colWidths addObject: [NSNumber numberWithFloat: rect.size.width]];
+            }
+        }
+
+      if (ci > 0 && [rowHeights count] == 0)
+        {
+          if (rect.size.width > 0.0)
+            {
+              [rowHeights addObject: [NSNumber numberWithFloat: rect.size.height]];
+            }
+        }
+    }
+  END_FOR_IN(_cells);
+
+  // Format the grid...
+  FOR_IN(NSGridCell*, c, _cells)
+    {
+      NSView *v = [c contentView];
+      NSUInteger ri = 0, ci = 0;
+          
+      // Get row and column index...
+      ci = i % num_col;
+      ri = i / num_col;
+      
+      // Get the row and col...
+      NSGridRow *row = [self rowAtIndex: ri];
+      NSGridColumn *col = [self columnAtIndex: ci];
+      NSRect rect = NSZeroRect;
+
+      if (v != nil)
+        {
+          rect = [v frame];
+        }
+      
+      // Do the math for the frame...
+      // NOTE: I am not sure why this is needed, the row and column heights are coming in via the nib
+      // as VERY small numbers (1.175... e-38) (see PR for this work, I have attached the XML nib there)
+      // so it is necessary to figure out the width of each column so that things can be properly spaced.
+      if (ci == 0)
+        {
+          current_y -= [[rowHeights objectAtIndex: ci] floatValue]; // rect.size.height;
+          current_x = 0.0;
+        }
+      
+      current_y -= [c yPlacement] - [row topPadding];
+      current_x += [c xPlacement] + [col leadingPadding] + [col width];
+      
+      if (v != nil)
+        {
+          rect.origin.x = current_x;
+          rect.origin.y = current_y;
+          [v setFrame: rect];
+          [self addSubview: v];
+        }
+      
+      current_x += [col trailingPadding] + [col width] + [[colWidths objectAtIndex: ci] floatValue] + _columnSpacing;
+      current_y -= [row bottomPadding] - [row height]; // - _rowSpacing; // add paddings after view...
+
+      // inc
+      i++;
+    }
+  END_FOR_IN(_cells)
+}
+
+- (instancetype) initWithFrame: (NSRect)frameRect
+{
+  self = [super initWithFrame: frameRect];
+
+  if (self != nil)
+    {
+      _rows = [[NSMutableArray alloc] init];
+      _columns = [[NSMutableArray alloc] init];
+      _cells = [[NSMutableArray alloc] init];
+    }
+
+  [self _refreshCells];
+  
+  return self;
+}
+
+- (instancetype) initWithViews: (NSArray *)rows // an array of arrays
+{
+  self = [self initWithFrame: NSZeroRect];
+
+  NSUInteger c = 0;
+  NSUInteger r = 0;
+
+  if (self != nil)
+    {
+      NSUInteger rc = [rows count], cc = 0;
+
+      if (rc > 0)
+        {
+          cc = [[rows objectAtIndex: 0] count];
+        }
+      
+      _cells = [[NSMutableArray alloc] initWithCapacity: rc * cc];
+      _rows = [[NSMutableArray alloc] initWithCapacity: rc];
+      _columns = [[NSMutableArray alloc] initWithCapacity: cc];
+
+      FOR_IN(NSArray*, row, rows)
```

Comment:
This whole code makes no sense to me. Could you please write some test code that exercises this method with more than one row and column?